lib/handler/gzip.c

```diff
@@ -0,0 +1,184 @@
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <zlib.h>
+#include "h2o.h"
+
+#define Z_GZIP_ENCODING 16
+#define MAX_BUF_SIZE 65000
+
+typedef struct st_gzip_encoder_t {
+    h2o_ostream_t super;
+    int wbits;
+    int memlevel;
+    z_stream zstream;
+    h2o_mem_pool_t *mempool;
+    H2O_VECTOR(h2o_iovec_t) outbufs;
+} gzip_encoder_t;
+
+static void *gzip_encoder_alloc(void *opaque, unsigned int items, unsigned int size)
+{
+    gzip_encoder_t *self = opaque;
+
+    /*
+     * I referenced from ngx_http_gzip_filter_module.c and write the code below (alloc len guard).
+     * Maybe this will help for memory alignment? IMHO The guard can be removed though.
+     */
+    unsigned int alloc = items * size;
+
+    if (alloc % 512 != 0 && alloc < 8192) {
+        alloc = 8192;
+    }
+
+    return h2o_mem_alloc_pool(self->mempool, alloc);
+}
+
+static void gzip_encoder_free(void *opaque, void *address)
+{
+    /* we use mempool here, so no need of free */
+}
+
+static void send_gzip(h2o_ostream_t *_self, h2o_req_t *req, h2o_iovec_t *inbufs, size_t inbufcnt, int is_final)
+{
+    gzip_encoder_t *self = (void *)_self;
+    
+    /*
+     * Well... The code below looks strange... I first tried to do gzip deflate in this way:
+     * Each time this function called, a buffer for output is allocated for all the input passed.
+     * If this call is not the final one, I just wait for the next call and do not pass the output
+     * to the next filter since I need to change Content-Length when the whole response has been
+     * deflated. So I don't need chunked filter anymore. Then the problem comes that I cannot get
+     * the next call since I don't pass the output to the next filter, the callback for sending
+     * next part of the response cannout be called...
```

Comment:
What you need to do here is:
* allocate outbufs if it is not allocated yet, or reuse the outbufs
* emit compressed data to outbufs, and pass it to the next stream (you may need to grow the outbufs while performing compression)

It is guaranteed that send_gzip will not be called until all the data sent to the next filter is sent to the client; therefore you can simply reuse the buffer when `send_gzip` is called again.